### PR TITLE
Feature/alpha/rbac

### PR DESCRIPTION
**Pull request by Ryanmtate**
WIP: opening as a draft for early feedback and discussion.

For this initial version, I've opted for using a basic `role`, `object`, `action` based RBAC policy, where the role is either `root` or `default`.

## Basic Types

```rust
// Policy Mapping
pub type Policy = HashMap<Role, HashMap<Object, Vec<Action>>>;


// Authorization tuple to find enforcement policy;
pub type AuthRequest = (Role, Object, Action);
```

## Auth Trait

```rust

#[async_trait]
pub trait Auth {    
    async fn enforce(&self, request: AuthRequest, role: Role) -> Result<bool, AuthError>;
    async fn set_role(&mut self, role: Role) -> Result<(), AuthError>;
    fn get_role(&self) -> Option<&Role>;
}
```

## Basic Implementation of Auth Trait

```rust
#[async_trait]
impl Auth for BasicAuth {
    async fn set_role(&mut self, role: Role) -> Result<(), AuthError> {
        self.role = Some(role);
        Ok(())
    }

    fn get_role(&self) -> Option<&Role> {
        self.role.as_ref()
    }

    async fn enforce(&self, request: AuthRequest, role: Role) -> Result<bool, AuthError> {

        let is_role_match = request.0 == role;
        
        let is_allowed = self
            .policy
            .get(&request.0)
            .map(|role| {
                role
                    .get(&request.1)
                    .map(|actions| actions.iter().any(|action| action == &request.2))
                    .unwrap_or(false)
            })
            .unwrap_or(false);

        Ok(is_role_match && is_allowed)
    }
}
```

## Use in API SharedContext

```rust

if let Ok(authorized) = ctx.authorize((Role::Root, Object::SpuGroup, Action::Create)).await {
        if !authorized {
            trace!("authorization failed for role: {:?}", ctx.role());
            // ... handle auth error;
        }
    };
```

### CODE REVIEW

File: src/auth/src/abac.rs [hunk 1]
```diff
@@ -0,0 +1,33 @@
+use super::{Auth, AuthRequest, Role, Object, Action};
```
Comment (sehz) [inline src/auth/src/abac.rs:1]: don't need for now.

File: src/auth/src/auth.rs [hunk 2]
```diff
@@ -0,0 +1,49 @@
+use crate::error::AuthorizationError;
+
+use async_trait::async_trait;
+use fluvio_service::IdentityContext;
+use fluvio_stream_model::core::Spec;
+use serde::{Serialize, Deserialize};
+
+
+#[derive(Eq, PartialEq, Hash, Serialize, Deserialize, Clone, Debug)]
+pub enum Role {
```
Comment (sehz) [inline src/auth/src/auth.rs:10]: don't need hard wired role now

File: src/auth/src/basic.rs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+impl BasicAuth {
+    pub fn new() -> Self {
+        BasicAuth {
+            role: None,
+            policy: BasicAuth::policy(),
+        }
+    }
+
+    fn policy() -> Policy {
```
Comment (sehz) [inline src/auth/src/basic.rs:23]: only root policy should be defined

File: src/auth/src/basic.rs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        policy.insert(Role::Default, default_policy);
+
+        // Finally, return the policy
+        policy
+    }
+}
+
+#[async_trait]
+impl Auth for BasicAuth {
+    async fn enforce<I: IdentityContext + Sync>(&self, request: AuthRequest, identity: I) -> Result<bool, AuthError> {
```
Comment (sehz) [inline src/auth/src/basic.rs:64]: add unit test

File: src/auth/src/error.rs [hunk 5]
```diff
@@ -0,0 +1,10 @@
+use thiserror::Error;
+use std::io::Error as IoError;
+
+#[derive(Error, Debug)]
+pub enum AuthorizationError {
+    #[error("io error")]
+    IoError(#[from] IoError),
+    #[error("permission denied")]
+    PermissionDenied,
```
Comment (sehz) [inline src/auth/src/error.rs:9]: permission deny is not error.  it should be just return value from enforcer

File: src/auth/src/rbac.rs [hunk 6]
```diff
@@ -0,0 +1,77 @@
+use super::{Authorization, AuthorizationRequest, Role, Object, Action, Policy};
```
Comment (sehz) [inline src/auth/src/rbac.rs:1]: duplicate of policy?

**Review by sehz — CHANGES_REQUESTED** [review]

File: src/auth/src/auth.rs (hunk 2 above)
Comment (Ryanmtate) [reply src/auth/src/auth.rs:10]: why would we not need a role if we're going to write the logic to enforce the policy in the fluvio code? @nacardin I thought we all agreed that this was set?

File: src/auth/src/abac.rs (hunk 1 above)
Comment (Ryanmtate) [reply src/auth/src/abac.rs:1]: This file is going to be deleted. Only `policy.rs` will exist. rbac, basic, and abac are to be removed. Spoke about this yesterday on the call.

File: src/auth/src/basic.rs (hunk 3 above)
Comment (Ryanmtate) [reply src/auth/src/basic.rs:23]: this file is to be deleted. See comment earlier.

File: src/auth/src/rbac.rs (hunk 6 above)
Comment (Ryanmtate) [reply src/auth/src/rbac.rs:1]: yep; to be deleted.

File: src/sc-schema/src/apis.rs [hunk 7]
```diff
@@ -12,6 +12,7 @@ use dataplane::derive::Decode;
 #[derive(Encode, Decode, PartialEq, Debug, Clone, Copy)]
 #[repr(u16)]
 pub enum AdminPublicApiKey {
+    Auth = 8,
```
Comment (Ryanmtate) [inline src/sc-schema/src/apis.rs:15]: remove

File: src/sc-schema/src/objects/auth.rs [hunk 8]
```diff
@@ -0,0 +1,22 @@
+#![allow(clippy::assign_op_pattern)]
```
Comment (Ryanmtate) [inline src/sc-schema/src/objects/auth.rs:1]: this file can be removed.

File: src/sc-schema/src/request.rs [hunk 9]
```diff
@@ -4,30 +4,29 @@
...
 use super::objects::*;
-
 use super::AdminPublicApiKey;
 
 #[derive(Debug, Encode)]
 pub enum AdminPublicRequest {
     // Mixed
     ApiVersionsRequest(RequestMessage<ApiVersionsRequest>),
-
+    AuthRequest(RequestMessage<AuthRequest>),
```
Comment (Ryanmtate) [inline src/sc-schema/src/request.rs:29]: remove.

File: src/sc-schema/src/request.rs [hunk 10]
```diff
@@ -56,6 +55,7 @@ impl ApiMessage for AdminPublicRequest {
             api_key
         );
         match api_key {
+            AdminPublicApiKey::Auth => api_decode!(Self, AuthRequest, src, header),
```
Comment (Ryanmtate) [inline src/sc-schema/src/request.rs:58]: remove.

File: src/auth/src/auth.rs (hunk 2 above)
Comment (nacardin) [reply src/auth/src/auth.rs:10]: That's what I thought... If not where would the role definitions come from? Config?

File: src/sc-schema/src/apis.rs (hunk 7 above)
Comment (Ryanmtate) [reply src/sc-schema/src/apis.rs:15]: put back

File: src/sc-schema/src/objects/auth.rs (hunk 8 above)
Comment (Ryanmtate) [reply src/sc-schema/src/objects/auth.rs:1]: put back

**Comment by sehz** [comment]
replaced by #420

**Pull request closed without merging**